<a href="https://colab.research.google.com/github/NikhilBHADHRAGIRI/My_files/blob/main/DAA1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Answer1:

After running the given bugged code the merge process is like as below

In [ ]:
from IPython.display import Image
Image(url="https://github.com/NikhilBHADHRAGIRI/My_files/blob/main/PICTURE1.jpeg?raw=true")

Bugged mergesort code:

In [ ]:
import math
def merge(arr, p, q, r):
    L = arr[p:q]
    R = arr[q:r+1]
    L.append(math.inf)
    R.append(math.inf)
    i, j = 0, 0
    for k in range(p, r+1):
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        
    
    
def mergeSort(arr, p, r,):
    if p < r:
        q = (p+r) // 2
        
        mergeSort(arr, p, q)
        
        mergeSort(arr, q+1, r)
        
        merge(arr, p, q, r)
        
        
    return arr
arr =[5,4,9,1]
print(arr)
p=0
r=len(arr)-1
print(mergeSort(arr,p,r,))

[5, 4, 9, 1]
[4, 5, 9, 1]


Here the bugged code is giving an output as unsorted array.The bug in the code lies in the merge function.
L = arr[p:q] should be changed to L = arr[p:q+1] and
R = arr[q:r+1] should be changed to R = arr[q+1:r+1].
Now after debugging , the merge process will be like

In [ ]:
from IPython.display import Image
Image(url="https://github.com/NikhilBHADHRAGIRI/My_files/blob/main/PICTURE2.jpeg?raw=true")

Debugged Code:

In [ ]:
import math
def merge(arr, p, q, r):
    L = arr[p:q+1]
    R = arr[q+1:r+1]
    L.append(math.inf)
    R.append(math.inf)
    i, j = 0, 0
    for k in range(p, r+1):
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        
    
    
def mergeSort(arr, p, r,):
    if p < r:
        q = (p+r) // 2
        
        mergeSort(arr, p, q)
        
        mergeSort(arr, q+1, r)
        
        merge(arr, p, q, r)
        
        
    return arr
arr =[5,4,9,1]
print(arr)
p=0
r=len(arr)-1
print(mergeSort(arr,p,r,))

[5, 4, 9, 1]
[1, 4, 5, 9]


The result of thid debugged code is a sorted array.

Answer2.1:

Given array is arr = [x for x in range(1<< 10, 0, -1)]
If we pass this array into an insertion sort code

Insertion Sort code:

In [ ]:
arr = [x for x in range(1<<10, 0, -1)]
l=len(arr)
c=0
for j in range(1,l):
    k=arr[j]
    i=j-1
    while(i>=0 and arr[i]>k):
        arr[i+1]=arr[i]
        i=i-1
        c=c+1
    arr[i+1]=k
print(c)

523776


The total number of iterations the inner while loop
executes is 523776.

Answer 2.2:

If we pass the array arr = [x for x in range(1<< 10, 0, -1)] into merge sort,then

Merge Sort code:

In [ ]:
import math
def merge(arr, p, q, r):
    L = arr[p:q+1]
    R = arr[q+1:r+1]
    L.append(math.inf)
    R.append(math.inf)
    i, j = 0, 0
    count = 0
    for k in range(p, r+1):
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        count=count+1
    return count
    
def mergeSort(arr, p, r,d):
    if p < r:
        q = (p+r) // 2
        
        mergeSort(arr, p, q,d)
        
        mergeSort(arr, q+1, r,d)
        
        x=merge(arr, p, q, r)
        
        d.append(x)
    return arr,d


d=[]
arr = [x for x in range(1<<10, 0, -1)]
p=0
r=len(arr)-1
mergeSort(arr,p,r,d)
c=sum(d)
print(c)

10240


The total number of iterations the for loop executes is 10240. 

Answer 2.3:

Given array is arr = [x for x in range(1<< 10, 0, -1)] If we pass this array into an insertion sort code

Insertion Sort code:

In [ ]:
arr = [x for x in range(1<<12, 0, -1)]
l=len(arr)
c=0
for j in range(1,l):
    k=arr[j]
    i=j-1
    while(i>=0 and arr[i]>k):
        arr[i+1]=arr[i]
        i=i-1
        c=c+1
    arr[i+1]=k
print(c)

8386560


The total number of iterations the inner while loop executes is 8386560.

If we pass the array arr = [x for x in range(1<< 10, 0, -1)] into merge sort,then

Merge Sort code:

In [ ]:
import math
def merge(arr, p, q, r):
    L = arr[p:q+1]
    R = arr[q+1:r+1]
    L.append(math.inf)
    R.append(math.inf)
    i, j = 0, 0
    count = 0
    for k in range(p, r+1):
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        count=count+1
    return count
    
def mergeSort(arr, p, r,d):
    if p < r:
        q = (p+r) // 2
        
        mergeSort(arr, p, q,d)
        
        mergeSort(arr, q+1, r,d)
        
        x=merge(arr, p, q, r)
        
        d.append(x)
    return arr,d


d=[]
arr = [x for x in range(1<<12, 0, -1)]

p=0
r=len(arr)-1
mergeSort(arr,p,r,d)
coun=sum(d)
print(coun)
print(len(arr))

49152
4096


The total number of iterations the for loop executes is 49152.

For arr= [x for x in range(1<<10, 0, -1)],n=1024,the loop in insertion sort executes 523776 times which is approximately n^2/2 which defines its time complexity O(n^2) and the loop in merge sort executes 10240 times which defines its time complexity O(nlog(n)).Similarly with the othe array as well.This means that merge sort is more efficient sorting algorithm than insertion sort. 

Answer 3:


Bubble Sort code:

In [26]:
def Bubblesort(Arr):
  l=len(Arr)
  for i in range(l):
    for j in range(l-1-i):
      if Arr[j] > Arr[j + 1]:
        t=Arr[j]
        Arr[j]=Arr[j+1]
        Arr[j+1]=t
  return(Arr)
Arr=[5,4,7,6]
print(Bubblesort(Arr))

[4, 5, 6, 7]


In order to demonstrate that Bubble-Sort is accurate, we must demonstrate that the end array Arr' is a permutation (reordering) of the initial array Arr such that Arr'[1] Arr'[2] ... Arr'[n]. Because the procedure never adds any new elements, the array Arr must always be a permutation of the initial array.

It doesn't add new items to the array or remove old ones; it just swaps out the ones that are already there. As a result, Arr' is a permutation of Arr. We now need to demonstrate that Arr'[1] Arr'[2] ... Arr'[n].
First, we establish that the inner for loop on lines 2-4 of Bubble-Sort satisfies the following

Loop invariant: Prior to each inner for loop iteration, the subarray Arr[i...Arr.length]'s value is contained in Arr[i...j] (at least one of them does, in the case of a tie).
Initialization: Prior to the first iteration of the loop, we must demonstrate that the loop invariant is true. Before

j = Arr.length for the initial iteration of the loop. We must thus demonstrate that the subarray Arr[i...Arr.length]'s minimal value is contained within it. Simply put, this is accurate. an invariant loop The smallest value in the subarray Arr[i...Arr.length] occurs within Arr[i...j] prior to any given inner for loop iteration (at least one of them does, in the case of a tie).
Initialization: Prior to the loop's initial iteration, we must demonstrate that the loop's invariant still holds. j = Arr.length was the condition before to the loop's initial iteration. Therefore, we must demonstrate that the minimal value in the subarray Arr[i...Arr.length] occurs within Arr[i...Arr.length]. This is obviously true.
Maintenance: Assume that the smallest value in the subarray Arr[i...Arr.length] is inside Arr[i...j] prior to a specific iteration of the inner for loop. We must demonstrate that the loop invariant still holds before the start of the subsequent iteration. There are three situations to take into account:
Case 1: Arr[i...Arr.length]'s minimal value can be found inside Arr[i...j-2]. The smallest value then continues to exist within Arr[i...j-2] after the iteration because the exchange step has no effect on it.
Case 2: Arr[j-1] is the subarray Arr[i...Arr.length]'s lowest value. When Arr[j-1] Arr[j], the exchange step is skipped since Arr[j-1] is the minimum. As a result, after iteration, the minimal value is still found in Arr[i...j-1].
Case 3: Arr[j] is the subarray Arr[i...Arr.length]'s lowest value. The minimum is now included within Arr[i...j-1] because Arr[j] Arr[j-1] since Arr[j] is the minimum. Since they have the same value, it doesn't matter if Arr[j] == Arr[j-1] because there is no exchange;the smallest value of Arr[i...Arr.length] nevertheless appears at least once within Arr[i...j-1]
Since in each of these scenarios the minimum value of Arr[i...Arr.length] occurs within Arr[i...j-1] following the iteration, the minimal value of Arr[i...Arr.length] occurs within Arr[i...j] at the start of the following iteration. Consequently, the loop invariant is valid for the following iteration.


On lines 1-4 of the outer for loop, we then demonstrate that the following loop invariant is valid:
an invariant loop The smallest i-1 elements are in the array's initial i-1 positions at the beginning of each iteration, ordered according to size.
Initialization: I was initially set to 1 before the first iteration. The statement "The smallest 0 elements are in the first 0 places in the array, in sorted order" is vacuously true.
Maintenance: Assume that, prior to a specific iteration of the loop, the smallest i-1 elements were arranged in the array's initial i-1 positions, according to their sequence of decreasing size. No items of the subarray can be changed by the inner for loop. Arr[1...i-1] since it only ever modifies elements Arr[j] and Arr[j-1], and across all j i+1 by the loop limits during iterations of the inner for loop. As a result, after the iteration is finished, the smallest i-1 elements are still located in the first i-1 positions of the array, in sorted order. hence, as
The smallest element in the subarray Arr[i...Arr.length] is the ith smallest element overall. Arr must always be a permutation of the original array. Arr[i] is the lowest element of the subarray Arr[i...Arr.length] following iteration, according to the inner for loop invariant. This implies that the

The array is sorted so that the smallest I elements are located in the first I positions. The smallest i-1 elements are in the first i-1 positions in the array, sorted, as I increases at the beginning of the following iteration. Consequently, before the subsequent iteration, the loop invariant is still true.
Termination: After the end of the loop, i = Arr.length. So, the smallest Arr.length-1 elements of the
array are in the first Arr.length-1 places, in sorted order. This means that the whole array is in
sorted order, since the last element must be the maximum.


The array is in sorted order after the outer for loop, and the algorithm ends immediately after the outer for loop terminates, thus we can conclude that Arr'[1] Arr'[2] ... Arr'[n]

In order to demonstrate that the final array Arr' is a permutation (reordering) of the initial array Arr, we showed that Arr'[1] Arr'[2] ... Arr'[n] is the resulting sequence. The Sorting Problem is therefore solved, and Bubble-Sort is the right solution.